Add code to use encoders

ryanloeffelman opened the pull request:
Needs to be tested on working robot, to verify the speed is calculated
correctly, and accurately

Also updated a lot of names to match the Missouri S&T Robotics Competition Team standards

closes #35

Commit:
Add code to use encoders

Needs to be tested on working robot, to verify the speed is calculated
correctlly, and acurattly

Commit:
Fix spelling mistakes

Changed the desiered_speed variable to desired_speed

Commit:
Add custom msg to send encoder data back to ROS

Added code to Arduino to publish encoder ticks
Created a custom msg including Header, and Uint8

ryanloeffelman commented:
I know it currently doesn't work, I need help in getting the Arduino compiler to see the custom msg I created @islamelnabarawy

ryanloeffelman commented:
I think this link may help, http://wiki.ros.org/rosserial_client/Tutorials/Generating%20Message%20Header%20Files but I don't fully understand it

Commit:
Update comments to be more accurate

matthewia94 commented on ros_ws/src/control/CMakeLists.txt:
```diff
@@ -16,8 +17,17 @@ catkin_package(
+## Generate message in the 'msg' folder
+add_message_files(FILES Encoder.msg)
+
+
```
There is a missing function here which is what is causing the msg build failure.  Since this depends on std_msgs you need the line:

`generate_messages(DEPENDENCIES std_msgs)`

to come here.  This entire message block will also have to be moved up to be above the catkin_specific configuration block in order for it to build.

islamelnabarawy commented:
@ryanloeffelman @matthewia94 
Is this pull request ready for review or is it still waiting for a fix?

ryanloeffelman commented:
Matt fixed the main issue, however I have yet to push my changes, give me around 10ish minuets and it should be ready to review

Commit:
Fix issue with Arduino not seeing header

Allowed the Arduino to see the Ros Custom Header, will require remaking the
rosserial_python libraries after making, so that the arduino code can
 actully compile

ryanloeffelman commented:
I haven't actually been able to test the code on the Robot yet, as we have no encoders installed to test with, however the code will has at least been able to read the encoder successfully. The ROS message updating has yet to be tested. Also, I am starting to get a low memory warning on the Arduino Compiler, not to much of an issue, however If you see anything I do to save space, that would be great, thanks @islamelnabarawy

islamelnabarawy commented:
@ryanloeffelman The arduino file caused merge conflicts, and after resolving them I'm getting a compilation error saying it can't find Encoder.h. We might need to go over this together in person.

matthewia94 commented:
The arduino file needs you to build the message headers.  This is done with the make_libraries script contained in the rosserial_arduino package.  The section Install ros_lib into the Arduino Environment here: http://wiki.ros.org/rosserial_arduino/Tutorials/Arduino%20IDE%20Setup gives details on how to set it up

islamelnabarawy commented:
The error I'm getting while trying to compile the arduino code is:

motor.ino:14:29: fatal error: control/Encoder.h: No such file or directory

matthewia94 commented:
Correct, you would get this error if you have not run the make_libraries script as the message header is auto generated by ROS.  @ryanloeffelman Would you mind maybe adding the build process for custom messages to the arduino page on our wiki?

Basically you need to build the normal code and source, find the libraries folder for arduino, delete ros_lib directory if it exists and run:
`rosrun rosserial_arduino make_libraries.py .`

islamelnabarawy commented:
Ok, it's possible that I hadn't sourced the project before generating the libraries last time. This time it worked, but there was another error because the set_pwm_frequency() function needed to be forward-declared. I did that and it compiled. I'll add a comment to the relevant line of code.

ryanloeffelman commented:
At least its not just me who makes that mistake. Are you using the Arduino Compiler? I know its good practice to forward declare things, however Arduino doesn't usually care about that.

islamelnabarawy commented:
It might have been due to a merge error. I will have to merge the code manually to make sure the merge doesn't break anything, and I'll accept the pull request as soon as I'm done with that.

ryanloeffelman commented:
Sound good, however I wouldn't actually accept this pull request until we can test it on the Robot, which will be after the encoder have been successfully installed

islamelnabarawy commented:
Noted.

matthewia94 commented on ros_ws/src/arudino/motor/motor.ino:
```diff
@@ -73,6 +124,49 @@ ros::NodeHandle nodeHandle;
+    node_handle.advertise(encoderPub);
+
+    //Set the Inturupt on Pin 2
+    attachInterrupt(0, encoderCount, CHANGE);
```
Change to:
attachInterrupt(0, encoderCount, RISING);
This will update on the rising edge and we won't have to check if the change was from low to high or not.

matthewia94 commented on ros_ws/src/arudino/motor/motor.ino:
```diff
@@ -106,31 +200,63 @@ void velocityCallback(const std_msgs::Float64& msg)
+        encoder_ticks++; 
+    }
+    
+    encoder_pos_last = encoder_pos_current;
```
With the above change all that needs to be done in this function is to incremement encoder_ticks

matthewia94 replied:
Also this should take into account the direction the wheel is spinning and decrement if it is supposed to be spinning backwards.

matthewia94 commented on ros_ws/src/arudino/motor/motor.ino:
```diff
@@ -106,31 +200,63 @@ void velocityCallback(const std_msgs::Float64& msg)
+    encoderMessage.header.stamp = node_handle.now();
+    
+    //publish message
+    encoderPub.publish(&encoderMessage);
```
The encoder ticks should be reset to 0 in this function

Commit:
Fix recomended changes to attempt to fix encoders

Commit:
Merge branch 'master' of https://github.com/MST-Robotics/IGVC into encoder

Conflicts:
	ros_ws/src/arudino/motor/motor.ino

Commit:
Fix error i made while merging

Commit:
Update code to remove unecassarry functions

ryanloeffelman committed:
Update to attempt to fix encoder code

ryanloeffelman committed:
Renamed the misspelled Arduino Folder

ryanloeffelman committed:
Renamed that arduino folder

ryanloeffelman committed:
Update the Arudino Timing Functions to hopefully fix sync issues
